### DISCUSSION TITLE
An ifLet received a child action when child state was nil

### DISCUSSION THREAD

**Discussion by JaapWijnen** [pointfreeco/swift-composable-architecture#1663] [Q&A]
Hi! I'm having some trouble properly implementing some navigation. I've rebuilt the Inventory app from swiftui-navigation using TCA. 

When performing some actions in my app I get the An ifLet received a child action when child state was nil runtime error. I think it's caused by a binding reemitting an action before being deallocated.

When I press the add button to add a new item to the inventory enter a name and then press cancel the following happens:
- `.addButtonTapped` fires and sets the route so that the add item sheet is opened 
- `.setName` fires and changes the name of the item
- `.cancelButtonTapped` action is fired (this sets the route to nil)
- `.setName fires` (again?) with the same name as set by the first time it was fired. This causes the error since the route is already set to nil by the previous action. I'm not sure why the setName binding fires again and/or how to prevent it!  

I have a similar problem for a .forEach scoping where the following happens when I have an item in the inventory and press the delete button and then press Delete to confirm the deletion in the alert that's presented:
- `.deleteButtonTapped` fires and sets the route so that the delete alert is presented
- `.deleteConfirmationButtonTapped` fires after the confirmation is tapped. this removes the item from the inventory collection and therefore the route is deleted since it's contained in the `ItemRow.State` that is deleted from inventory
- `.setRoute` fires since the alert will disappear, this triggers the runtime error "forEach" received an action for a missing element. since the Element no longer exists in the array of `ItemRow`s

Help would be really appreciated!
Any tips? My code can be found here: https://github.com/JaapWijnen/swift-composable-architecture/tree/inventory-example

**Comment by mbrandonw** [comment]
Hi @JaapWijnen, this is unfortunately a known, long-standing bug in SwiftUI. Text field bindings fire twice for some reason. We've filed a [feedback](https://gist.github.com/mbrandonw/b4bdb52e7a71cde0d7024573d39e199c) about it, and there are been a few discussion about it in the past (e.g. [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/1093) and [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/800)).

There's a workaround to use a kind of `removeDuplicates()` method on bindings, and I think that's about the best you can do.

I'm not sure about the second one. I don't have enough time to dive into the code right now.

**Reply by JaapWijnen** [reply]
Thanks @mbrandonw that did the trick for the `.ifLet`
Hoping to fix the `.forEach` as well, maybe some of the upcoming tools will help me understand or someone has come across this issue as well :) Thanks for the help!

**Reply by JaapWijnen** [reply]
For some more context whoever might read this. 
The `InventoryFeature` contains an `IdentifiedArrayOf<ItemRow.State>`
and handles the removal of one of the items by the following logic in the reducer:
```swift
case .itemRow(let id, .deleteConfirmationButtonTapped):
            state.items.remove(id: id)
            return .none
```
The `ItemRow` reducer doesn't have any logic for that action. 
however that reducer does have a `.setRoute(Route?)` action that is used by a binding on a `.alert()` on the `ItemRowView`:
```swift
.alert(
    title: { Text(viewStore.name) },
    unwrapping: viewStore.binding(
        get: \.route,
        send: ItemRow.Action.setRoute
    ),
    case: /ItemRow.Route.deleteAlert,
    actions: { _ in
        Button("Delete", role: .destructive) {
            viewStore.send(.deleteConfirmationButtonTapped)
        }
    },
    message: { Text("Are you sure you want to delete this item?") }
)
```
So the item is deleted from the identified array by sending the `.deleteConfirmationButtonTapped` action. Therefore the item is missing from the array when this binding still emits a `.setRoute(nil)` action to the reducer. Which will then complain since the Inventory reducer and ItemRow reducer are chained as: 
```swift
_InventoryFeature()
    .forEach(\.items, action: /Action.itemRow) {
        ItemRow()
    }
```
